### PR TITLE
Make the backquote operator always return a string

### PR DESCRIPTION
**Pull request by knu**
This fixes this behavior:

```ruby
# Vanilla Ruby (on unix)
`nonexistentcommand` #=> Errno::ENOENT
`nonexistentcommand >/dev/null` == ""

# With ActiveSupport loaded
`nonexistentcommand` == nil # I believe this should be fixed
`nonexistentcommand >/dev/null` == "" # With an error message on stderr
```

With this PR merged:

```ruby
# With ActiveSupport loaded
`nonexistentcommand` == "" # With an error message on stderr
`nonexistentcommand >/dev/null` == "" # With an error message from the shell on stderr
```

### Summary

ActiveSupport overrides `` Kernel#` `` so that it would not raise
`Errno::ENOENT` but return `nil` instead if a given command were not
found.  In vanilla Ruby you can safely say `` `command`.strip `` or
`` `command`.empty? `` but with ActiveSupport loaded that's not the
case anymore.

A good function wrapper does not change the return value type, so the
core extension of `` Kernel#` `` should return an empty (mutable)
string when it decides to suppress an exception.

### CODE REVIEW

**Comment by matthewd** [comment]
Are we unifying a behaviour that's since been unified upstream in the other direction?

> ```ruby
> # Vanilla Ruby
> `nonexistentcommand` == ""
> ```

Everywhere I tried this just now I got an ENOENT, not a string. :confused:

**Comment by knu** [comment]
Oops, that description I added the last minute was wrong, I’ll fix it shortly.

**Comment by matthewd** [comment]
It sounds like maybe we should just remove the method. I don't think it was a strong opinion that it should fail quietly -- we just wanted it to do something consistently. If upstream is now meeting that condition, we don't need to do anything.

**Comment by knu** [comment]
@matthewd Sorry, I've fixed the description.

> It sounds like maybe we should just remove the method. I don't think it was a strong opinion that it should fail quietly -- we just wanted it to do something consistently. If upstream is now meeting that condition, we don't need to do anything.

That'd be an idea, but at least the upstream (Ruby) did not change anything with the said behavior.  `` `nonexistentcommand` `` will raise Errno::ENOENT at least on unix.

**Comment by matthewd** [comment]
> the upstream (Ruby) did not change anything with the said behavior

It clearly did, because the comment on this method says that [when it was written] on Unix, `` `badcmd` `` warned and returned nothing, and that's what it's trying to emulate. A quick experiment confirms:

```
irb(main):001:0> `nonexistentcommand`
(irb):1: command not found: nonexistentcommand
=> ""
irb(main):002:0> RUBY_VERSION
=> "1.8.7"
```

(Interestingly, an empty string, not a nil. I don't have 1.8.6 nearby to compare; maybe that was always inaccurate.)

I think this reinforces the idea that we should be removing it: we standardized on the 1.8.7 Unix behaviour, and now we're just (very) out of date.

**Comment by knu** [comment]
Hmm.  Is it really safe to break the assumption among Rails users that a backquote operator wouldn’t raise ENOENT?  I have no particular opinion and just think it is unexpectable and inappropriate for it to return nil.

**Comment by oniofchaos** [comment]
Anything stopping this from being merged?

**Comment by rafaelfranca** [comment]
I agree with Matthew that we should remove this method. @knu can you remove the method and add a CHANGELOG entry?

**Comment by knu** [comment]
@rafaelfranca Done.  Should I squash the commits?

File: activesupport/test/core_ext/kernel_test.rb [hunk 1]
```diff
@@ -1,9 +1,12 @@
 # frozen_string_literal: true
 
 require "abstract_unit"
+require "active_support/testing/stream"
 require "active_support/core_ext/kernel"
 
 class KernelTest < ActiveSupport::TestCase
+  include ActiveSupport::Testing::Stream
```
Comment (kamipo) [inline activesupport/test/core_ext/kernel_test.rb]: This has no longer used.

**Comment by knu** [comment]
Friendly ping.

**Comment by matthewd** [comment]
@knu sorry for the delay -- yes, please go ahead and squash. :+1:

@rafaelfranca do you have any thoughts on whether we should deprecate this first? I'm leaning towards treating it as a bugfix (though one we wouldn't backport) and merging this as is, but we _could_ put a deprecation inside the `rescue` first. Personally, I think the existing `STDERR.puts` should've been ugly enough to act as a deprecation -- that plus the fact we're aligning with standard ruby behaviour seems sufficient to me.

**Comment by knu** [comment]
@matthewd I've squashed all commits into one and updated the commit message.

I agree you can treat this as a mere bug fix!

**Comment by rafaelfranca** [comment]
I's also treat this as a bug fix. No need for deprecation.

**Comment by knu** [comment]
Someone?

**Comment by knu** [comment]
@y-yagi Done, and the CI build passed.

**Comment by y-yagi** [comment]
@knu Thanks!

**Pull request merged**